javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -1202,9 +1329,9 @@ private predicate onPath(DataFlow::Node nd, DataFlow::Configuration cfg, PathSum
   not cfg.isLabeledBarrier(nd, summary.getEndLabel())
   or
   exists(DataFlow::Node mid, PathSummary stepSummary |
-    reachableFromSource(nd, cfg, summary) and
+    onPath(mid, cfg, summary.append(stepSummary)) and
     flowStep(nd, id(cfg), mid, stepSummary) and
-    onPath(mid, id(cfg), summary.append(stepSummary))
+    reachableFromSource(nd, id(cfg), summary)
```

Comment:
I've had some troubles with bad join orders related to `summary.append(stepSummary)`, it might be the same thing you've tried to get around here. 

I think I fixed a similar issue by outlining a predicate (in 6f0e50724278a29687c16a4901ca9155a69e48a5).

Reply:
Yes, I'm aware of the outlining trick.